utils/common/collation.h

```diff
@@ -166,7 +166,11 @@ class Charset
   }
   int strnncollsp(const utils::ConstString& str1, const utils::ConstString& str2) const
   {
-    return mCharset->strnncollsp(str1.str(), str1.length(), str2.str(), str2.length());
+    // nullptr handling below should return values as if nulls are substituted with empty string.
+    // please note that ConstString has an assertion so that nullptr data has zero length.
+    const char* s1 = str1.str();
+    const char* s2 = str2.str();
+    return mCharset->strnncollsp(s1 ? s1 : "", str1.length(), s2 ? s2 : "" , str2.length());
```

Comment:
What is an empty string semantics here? Is it still NULL or what?

Reply:
If it is NULL then two NULLs should be equal according with the code that doesn't sound right to me.

Reply:
What if you apply NullString::eq() logic here?

Reply:
We should not.

strnncollsp should have the same semantics as strnncollsp from charsets in server, which does not know what SQL's NULLs are. The code we are talking about is a defensive programming.

Reply:
Hm, but this approach will return true for NULL == NULL, right?

Reply:
Right.

NULL's usualy are dealt with elsewhere. Typical use cases for strnncollsp are "compute ranges" and sorting. These use cases insulate strnncollsp from NULLs.

This thing is just wrapper over what server provides, coded in "defensive style".